dot/state/grandpa.go

```diff
@@ -24,44 +26,570 @@ var (
 	currentSetIDKey   = []byte("setID")
 )
 
+var (
+	errPendingScheduledChanges = errors.New("pending scheduled changes needs to be applied")
+	errDuplicateHashes         = errors.New("duplicated hashes")
+	errAlreadyHasForcedChanges = errors.New("already has a forced change")
+	errUnfinalizedAncestor     = errors.New("ancestor with changes not applied")
+	errLowerThanBestFinalized  = errors.New("current finalized is lower than best finalized header")
+
+	ErrNoChanges = errors.New("cannot get the next authority change block number")
+)
+
+type pendingChange struct {
+	bestFinalizedNumber uint32
+	delay               uint32
+	nextAuthorities     []types.Authority
+	announcingHeader    *types.Header
+}
+
+func (p pendingChange) String() string {
+	return fmt.Sprintf("announcing header: %s (%d), delay: %d, next authorities: %d",
+		p.announcingHeader.Hash(), p.announcingHeader.Number, p.delay, len(p.nextAuthorities))
+}
+
+func (p *pendingChange) effectiveNumber() uint {
+	return p.announcingHeader.Number + uint(p.delay)
+}
+
+// appliedBefore compares the effective number between two pending changes
+// and returns true if:
+// - the current pending change is applied before the target
+// - the target is nil and the current contains a value
+func (p *pendingChange) appliedBefore(target *pendingChange) bool {
+	if p != nil && target != nil {
+		return p.effectiveNumber() < target.effectiveNumber()
+	}
+
+	return p != nil
+}
+
+type isDescendantOfFunc func(parent, child common.Hash) (bool, error)
+
+type pendingChangeNode struct {
+	change *pendingChange
+	nodes  []*pendingChangeNode
+}
+
+func (c *pendingChangeNode) importScheduledChange(blockHash common.Hash, blockNumber uint, pendingChange *pendingChange,
+	isDescendantOf isDescendantOfFunc) (imported bool, err error) {
+	announcingHash := c.change.announcingHeader.Hash()
+
+	if blockHash.Equal(announcingHash) {
+		return false, errDuplicateHashes
+	}
+
```

Comment:
what happens if there's two scheduled changes per block?

Reply:
We consider just the first one the other ones will be dropped. This check is similar to the one made by the `ForkTree` struct when importing the data for a block